File: 211-design-add-and-search-words-data-structure/solution.cpp
#include <string>
#include <unordered_map>

struct TrieNode
{
public:
  std::unordered_map<char, TrieNode *> edges;
};

class WordDictionary
{
private:
  TrieNode *root = new TrieNode();

  static bool search_helper(const size_t index, const TrieNode *const node, const std::string &word)
  {
    if (node == nullptr)
    {
      return false;
    }

    if (index == word.size())
    {
      return node->edges.contains('\0');
    }

    const char c = word[index];
    if (c == '.')
    {
      for (const auto &it : node->edges)
      {
        if (WordDictionary::search_helper(index + 1, it.second, word))
        {
          return true;
        }
      }

      return false;
    }

    const auto it = node->edges.find(c);
    if (it == node->edges.end())
    {
      return false;
    }

    return WordDictionary::search_helper(index + 1, it->second, word);
  }

public:
  WordDictionary() {}

  void addWord(std::string word)
  {
    auto node = this->root;
    for (const char c : word)
    {
      auto it = node->edges.find(c);
      if (it == node->edges.end() || it->second == nullptr)
      {
        auto result = node->edges.emplace(c, new TrieNode());
        it = result.first;
      }

      node = it->second;
    }

    node->edges.emplace('\0', nullptr);
  }

  bool search(std::string word)
  {
    return WordDictionary::search_helper(0, this->root, word);
  }
};

File: 211-design-add-and-search-words-data-structure/solution.py
from typing import Optional


class TrieNode:
  def __init__(self) -> None:
    self.edges: dict[str, Optional[TrieNode]] = {}


class WordDictionary:
  def __init__(self):
    self.root = TrieNode()

  def addWord(self, word: str) -> None:
    node = self.root
    for char in word:
      if char not in node.edges or node.edges[char] is None:
        node.edges[char] = TrieNode()

      node = node.edges[char]

    node.edges['\0'] = None

  def search(self, word: str) -> bool:
    def search_helper(index: int, node: Optional[TrieNode]) -> bool:
      if node is None:
        return False

      if index == len(word):
        return '\0' in node.edges

      char = word[index]
      if char == '.':
        for next_node in node.edges.values():
          if search_helper(index + 1, next_node):
            return True

        return False

      if char not in node.edges:
        return False

      return search_helper(index + 1, node.edges[char])

    return search_helper(0, self.root)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
